### DISCUSSION TITLE
Is it possible to run pact mock server without running a test?

### DISCUSSION THREAD

**Discussion by dormeiri** [pact-foundation/pact-js#887] [Q&A]
Hi all :)

Currently, we are running our services locally in isolation by binding the application interfaces to mocks using IoC.

For example:
```ts
interface OrdersServiceClient {
    getOrder(id: string): Promise<Order>;
}

// Mocked concrete implementation
class OrdersServiceClientMock implements OrdersServiceClient {
    private mockData: Order[] = [
        {
            id: 1,
            name: 'a'
        }, 
        {
            id: 2,
            name: 'b'
        }
    ];

    getOrder(id: string): Promise<Order> {
        return Promise.resolve(this.mockData.find((order) => order.id === id);
    }
}

// Live concrete implementation
class OrdersServiceClientImpl implements OrdersServiceClient {
    getOrder(id: string): Promise<Order> {
        return axios.get(`${ordersServiceSettings.baseUrl}/orders/${id}`);
    }
}

// IoC binding
container.bind(TYPE.ORDERS_SERVICE_CLIENT).to(ordersServiceSettings.useMock ? OrdersServiceClientMock : OrdersServiceClientImpl);
```

Now, if we want to start using Pact we will have two mock systems, one for tests, and one for an isolated local run.

I want to be able to configure and run the Pact mock server locally when running the application.
--or--
Use the existing mocked implementations to configure the expected interactions in the tests.

Is that possible? Do you have any idea on how to do that correctly?

--P.S

I thought that maybe we can implement a third concrete implementation that will implement each function as follows:

```ts
// Pact concrete implementation
class OrdersServiceClientPact implements OrdersServiceClient {
    private mockClient = new OrdersServiceClientMock();
    private liveClient = new OrdersServiceClientImpl();

    async getOrder(id: string): Promise<Order> {
        const expected = await this.mockClient.getOrder(id);
        // Add interaction, the response should be like `expected`
        return await this.liveClient.getOrder(id);
    }
}

// IoC binding
container.bind(TYPE.ORDERS_SERVICE_CLIENT).to(ordersServiceSettings.usePact ? OrdersServiceClientPact : (ordersServiceSetting.useMock ? OrdersServiceClientMock : OrdersServiceClientImpl));
```

This "Pact" concrete implementation will be bounded to the interface only while running tests. While I can still use the mocked concrete implementation while running the application locally.

This feels a bit complicated. Do you have a better suggestion?

**Comment by portyu9** [comment]
Hi @dormeiri 

Yes, but I would separate the two use cases.

For **contract tests**, don't create a third `OrdersServiceClientPact` implementation that defines interactions internally.

Your Pact test should exercise the **real** client:

```text
OrdersServiceClientImpl
        ↓
Pact mock server
```

rather than:

```text
OrdersServiceClientPact
        ↓
defines its own expected interactions
```

Otherwise the contract starts being generated from test/mocking code instead of proving what the real consumer actually sends.

A normal consumer Pact test should look conceptually like:

```js
await provider
    .addInteraction()
    // configure expected request/response
    .executeTest(async (mockServer) => {
        const client = new OrdersServiceClientImpl({
            baseUrl: mockServer.url
        })

        const order = await client.getOrder('1')

        // assertions
    })
```

For **running your application locally with a fake provider**, Pact now has a better tool for exactly that use case: the Pact Stub Server.

First generate your pact files from the consumer contract tests, then run them as a local API stub:

```bash
pact stub \
  --dir ./pacts \
  --port 8080 \
  --watch
```

or with the standalone executable:

```bash
pact-stub-server \
  --dir ./pacts \
  --port 8080 \
  --watch
```

Then configure your application locally to use:

```text
http://localhost:8080
```

as the provider URL.

This gives you:

```text
Consumer contract tests
        ↓
generate pact files
        ↓
Pact Stub Server
        ↓
local application can use those contracts as API stubs
```

So you don't necessarily need two independently maintained sets of mocked responses.

That said, Pact's stub server is intentionally fairly simple. If your existing local mocks have advanced behaviour/state that Pact files cannot conveniently model, keeping those local mocks is also completely reasonable.

The important part is that your Pact consumer tests should still call the **real `OrdersServiceClientImpl`**, because that is what makes the generated contract trustworthy.

Please mark this answer as accepted if it helped, thank you.
